monitoring/opencensus/main.go

```diff
@@ -0,0 +1,127 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+// [START monitoring_sli_metrics_opencensus_setup]
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"strconv"
+	"time"
+
+	"contrib.go.opencensus.io/exporter/stackdriver"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"golang.org/x/exp/rand"
+)
+
+// [END monitoring_sli_metrics_opencensus_setup]
+// [START monitoring_sli_metrics_opencensus_measure]
+// set up metrics
+var (
+	requestCount       = stats.Int64("oc_request_count", "total request count", "requests")
+	failedRequestCount = stats.Int64("oc_failed_request_count", "count of failed requests", "requests")
+	responseLatency    = stats.Float64("oc_latency_distribution", "distribution of response latencies", "s")
+)
+
+// [END monitoring_sli_metrics_opencensus_measure]
+
+// set up views
+// [START monitoring_sli_metrics_opencensus_view]
+var (
+	requestCountView = &view.View{
+		Name:        "oc_request_count",
+		Measure:     requestCount,
+		Description: "total request count",
+		Aggregation: view.Count(),
+	}
+	failedRequestCountView = &view.View{
+		Name:        "oc_failed_request_count",
+		Measure:     failedRequestCount,
+		Description: "count of failed requests",
+		Aggregation: view.Count(),
+	}
+	responseLatencyView = &view.View{
+		Name:        "oc_response_latency",
+		Measure:     responseLatency,
+		Description: "The distribution of the latencies",
+		// bucket definitions must be explicit
+		Aggregation: view.Distribution(0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000),
+	}
+)
+
+// [END monitoring_sli_metrics_opencensus_view]
+
+func main() {
+
+	// set up project ID
+	projectID := flag.String("project_id", "", "Cloud Project ID")
+	flag.Parse()
+
+	// register the views
+	if err := view.Register(requestCountView, failedRequestCountView, responseLatencyView); err != nil {
+		log.Fatalf("Failed to register the views: %v", err)
+	}
+	// [START monitoring_sli_metrics_opencensus_exporter]
+	// set up Cloud Monitoring exporter
+	sd, err := stackdriver.NewExporter(stackdriver.Options{
+		ProjectID:         *projectID,
+		MetricPrefix:      "opencensus-demo",
+		ReportingInterval: 60 * time.Second,
+	})
+	if err != nil {
+		log.Fatalf("Failed to create the Cloud Monitoring exporter: %v", err)
+	}
+	defer sd.Flush()
+	// Start the metrics exporter
+	sd.StartMetricsExporter()
+	defer sd.StopMetricsExporter()
+	// [END monitoring_sli_metrics_opencensus_exporter]
+	http.HandleFunc("/", handle)
+	log.Fatal(http.ListenAndServe(":8080", nil))
+}
+
+func handle(w http.ResponseWriter, r *http.Request) {
+	// create context
+	ctx, _ := tag.New(context.Background())
```

Comment:
Use `r.Context` to pick up anything coming though the request.

```suggestion
	ctx, _ := tag.New(r.Context())
```

Reply:
Fixed, thank you!